File: x86_trace_generator/pin/source/tools/ManualExamples/replacesigprobed.cpp
/*
 * Copyright (C) 2006-2021 Intel Corporation.
 * SPDX-License-Identifier: MIT
 */

//  Replace an original function with a custom function defined in the tool using
//  probes.  The replacement function has a different signature from that of the
//  original replaced function.

#include "pin.H"
#include <iostream>







typedef VOID* (*FP_MALLOC)(size_t);

// This is the replacement routine.
//
VOID* NewMalloc(FP_MALLOC orgFuncptr, UINT32 arg0, ADDRINT returnIp)
{
    // Normally one would do something more interesting with this data.
    //
    std::cout << "NewMalloc (" << std::hex << ADDRINT(orgFuncptr) << ", " << std::dec << arg0 << ", " << std::hex << returnIp << ")" << std::endl << std::flush;

    // Call the relocated entry point of the original (replaced) routine.
    //
    VOID* v = orgFuncptr(arg0);

    return v;
}

// Pin calls this function every time a new img is loaded.
// It is best to do probe replacement when the image is loaded,
// because only one thread knows about the image at this time.
//
VOID ImageLoad(IMG img, VOID* v)
{
    // See if malloc() is present in the image.  If so, replace it.
    //
    RTN rtn = RTN_FindByName(img, "malloc");

    if (RTN_Valid(rtn))
    {
        if (RTN_IsSafeForProbedReplacement(rtn))
        {
            std::cout << "Replacing malloc in " << IMG_Name(img) << std::endl;

            // Define a function prototype that describes the application routine
            // that will be replaced.
            //
            PROTO proto_malloc = PROTO_Allocate(PIN_PARG(void*), CALLINGSTD_DEFAULT, "malloc", PIN_PARG(int), PIN_PARG_END());

            // Replace the application routine with the replacement function.
            // Additional arguments have been added to the replacement routine.
            //
            RTN_ReplaceSignatureProbed(rtn, AFUNPTR(NewMalloc), IARG_PROTOTYPE, proto_malloc, IARG_ORIG_FUNCPTR,
                                       IARG_FUNCARG_ENTRYPOINT_VALUE, 0, IARG_RETURN_IP, IARG_END);

            // Free the function prototype.
            //
            PROTO_Free(proto_malloc);
        }
        else
        {
            std::cout << "Skip replacing malloc in " << IMG_Name(img) << " since it is not safe." << std::endl;
        }
    }
}

/* ===================================================================== */
/* Print Help Message                                                    */
/* ===================================================================== */

INT32 Usage()
{
    std::cerr << "This tool demonstrates how to replace an original" << std::endl;
    std::cerr << " function with a custom function defined in the tool " << std::endl;
    std::cerr << " using probes.  The replacement function has a different " << std::endl;
    std::cerr << " signature from that of the original replaced function." << std::endl;
    std::cerr << std::endl << KNOB_BASE::StringKnobSummary() << std::endl;
    return -1;
}

/* ===================================================================== */
/* Main: Initialize and start Pin in Probe mode.                         */
/* ===================================================================== */

int main(INT32 argc, CHAR* argv[])
{
    // Initialize symbol processing
    //
    PIN_InitSymbols();

    // Initialize pin
    //
    if (PIN_Init(argc, argv)) return Usage();

    // Register ImageLoad to be called when an image is loaded
    //
    IMG_AddInstrumentFunction(ImageLoad, 0);

    // Start the program in probe mode, never returns
    //
    PIN_StartProgramProbed();

    return 0;
}

File: x86_trace_generator/pin/bindings/glue/pinglueserver/tests/Scripts/replacesigprobed.py
#
# Copyright (C) 2024-2024 Intel Corporation.
# SPDX-License-Identifier: MIT
#

import pin

total = 0

info = open("script_replacesigprobed.out", "w")

def Probe_Exit_IA32(orgFuncptr, proto, code):
    global total
    info.write(f"In Probe_Exit_IA32, total calls to NewMalloc:{total}\n")
    info.close()
    
    return pin.GLUE_CallApplicationFunctionProbed(orgFuncptr, proto, code)
	
def NewMalloc(orgFuncptr, proto, arg0, returnIp):
    global total
    total += 1
    info.write(f"NewMalloc ({orgFuncptr}, {arg0}, {hex(returnIp)})\n")

    return pin.GLUE_CallApplicationFunctionProbed(orgFuncptr, proto, arg0)

def image_instrumentation_cb(img):
    rtn = pin.RTN_FindByName(img, "malloc")

    if pin.RTN_Valid(rtn):
        if pin.RTN_IsSafeForProbedReplacement(rtn):
            info.write(f"Replacing malloc in {pin.IMG_Name(img)}\n")
            proto_malloc = pin.PROTO_Allocate(pin.PIN_PARG_POINTER, pin.CALLINGSTD_DEFAULT, "malloc", pin.PIN_PARG_INT)

            pin.RTN_ReplaceSignatureProbed(rtn, NewMalloc, proto_malloc, pin.IARG_ORIG_FUNCPTR, 
                                           pin.IARG_PTR, proto_malloc, pin.IARG_FUNCARG_ENTRYPOINT_VALUE, 0, pin.IARG_RETURN_IP)

            # In case the proto is also passed to the replacement function (as IARG_PTR) for GLUE_CallApplicationFunction
            # We can NOT free it here.
            # pin.PROTO_Free(proto_malloc)
        else:
            info.write(f"Skip replacing malloc in {pin.IMG_Name(img)} since it is not safe.\n")
            
    exitRtn = pin.RTN_FindByName(img, "_exit")
    if pin.RTN_Valid(exitRtn) and pin.RTN_IsSafeForProbedReplacement(exitRtn):
        proto_exit = pin.PROTO_Allocate(pin.PIN_PARG_VOID, pin.CALLINGSTD_DEFAULT, "_exit", pin.PIN_PARG_INT)

        pin.RTN_ReplaceSignatureProbed(exitRtn, Probe_Exit_IA32, proto_exit, 
                                       pin.IARG_ORIG_FUNCPTR, pin.IARG_PTR, proto_exit,
                                       pin.IARG_FUNCARG_ENTRYPOINT_VALUE, 0)
        info.write(f"Replaced exit() in:{pin.IMG_Name(img)}\n")

pin.IMG_AddInstrumentFunction(image_instrumentation_cb)
